work/worker.go

```diff
@@ -829,6 +887,60 @@ func (env *Task) commitTransaction(tx *types.Transaction, bc BlockChain, rewardb
 	return nil, receipt.Logs
 }
 
+func (env *Task) commitBundleTransaction(bundle *builder.Bundle, bc BlockChain, rewardbase common.Address, vmConfig *vm.Config) (error, *types.Transaction, []*types.Log) {
+	lastSnapshot := env.state.Copy()
+	gasUsedSnapshot := env.header.GasUsed
+	txs := []*types.Transaction{}
+	receipts := []*types.Receipt{}
+	logs := []*types.Log{}
+
+	for _, txOrGen := range bundle.BundleTxs {
+		var tx *types.Transaction
+		if gen, ok := txOrGen.(builder.TxGenerator); ok {
+			var err error
+			tx, err = gen(env.state.GetNonce(rewardbase) + 1)
+			if err != nil {
+				for _, txInBundle := range bundle.BundleTxs {
+					switch v := txInBundle.(type) {
+					case *types.Transaction:
+						v.MarkUnexecutable(true)
+					}
+				}
+				logger.Error("TxGenerator error", "error", err)
+				*env.state = *lastSnapshot
+				env.header.GasUsed = gasUsedSnapshot
+				return err, &types.Transaction{}, nil
+			}
+		} else {
+			tx = txOrGen.(*types.Transaction)
+		}
+
+		receipt, _, err := bc.ApplyTransaction(env.config, &rewardbase, env.state, env.header, tx, &env.header.GasUsed, vmConfig)
+		if err != nil {
+			if err != vm.ErrInsufficientBalance && err != vm.ErrTotalTimeLimitReached {
+				for _, txInBundle := range bundle.BundleTxs {
+					switch v := txInBundle.(type) {
+					case *types.Transaction:
+						v.MarkUnexecutable(true)
+					}
+				}
+			}
+			*env.state = *lastSnapshot
+			env.header.GasUsed = gasUsedSnapshot
```

Comment:
AFAIK we need to roll back the state if the tx is *reverted*.

The error returned by `ApplyTransaction` isn't a VM error, but a validation or time limit error. We also need to check the `receipt.status` if it has any vmerr.

Reply:
If it is vmerr does that mean there is no need to roll back?

Reply:
No, I think vmerr should lead to a rollback. 

In [LendTx, ApprovalTx, SwapTx], if `SwapTx` is reverted but not rollback whole bundle, the proposer can't receive KAIA spent in `LendTx`. I thought it's the main purpose of bundling. cc. @ian0371 @blukat29

Reply:
This is an extension of `commitTransaction`, in which state is rolled back (via `RevertToSnapshot`) when EVM reverts.
Likewise, I think this can handle EVM error.

Reply:
https://github.com/kaiachain/kaia/blob/cf0ebf012cf7d60bc67072836dd2d9c0cd05f59a/blockchain/state_transition.go#L461
I think Lewis also has a valid point.
Isn't the state rollback performed in `commitTransaction`'s `RevertToSnapshot` upon EVM revert?

Reply:
I also agree with the point Lewis mentioned.

> Isn't the state rollback performed in commitTransaction's RevertToSnapshot upon EVM revert?

Yes, in the case of vmerr, commitTransaction's revert is not executed.

Reply:
State rollback: `lastSnapshot` is a copy of statedb before bundle execution, so any state changes are discarded.
Txs discard: txs in bundle is added to `env.txs` in the outer loop, which means only when all txs succeeded.

Reply:
@ian0371 
As you said, there is a difference in the scope of rollback when vmerr fails as a single tx and when vmerr fails as a bundle.
Can we tolerate this difference?

Reply:
I'm in favor of tolerating this and rolling back anyway in the case of vmerr.

Reply:
In the case of commitBundleTransaction, we have changed it so that a revert occurs if receipt.Status is not Successful. PTAL 582fd7c